pkg/cmd/admin/migrate/storage/storage.go

```diff
@@ -0,0 +1,431 @@
+	options := &MigrateAPIStorageOptions{
+		ResourceOptions: migrate.ResourceOptions{
+			In:      in,
+			Out:     out,
+			ErrOut:  errout,
+			Include: []string{"imagestream", "image", "secrets"},
+		},
+	}
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s REGISTRY/NAME=REGISTRY/NAME [...]", name),
+		Short:   "Update the stored version of API objects",
+		Long:    internalMigrateStorageLong,
+		Example: fmt.Sprintf(internalMigrateStorageExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(options.Complete(f, cmd, args))
+			kcmdutil.CheckErr(options.Validate())
+			if err := options.Run(); err != nil {
+				// TODO: move met to kcmdutil
+				if err == cmdutil.ErrExit {
+					os.Exit(1)
+				}
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+	options.ResourceOptions.Bind(cmd)
+
+	return cmd
+}
+
+func (o *MigrateAPIStorageOptions) Complete(f *clientcmd.Factory, c *cobra.Command, args []string) error {
+	o.ResourceOptions.SaveFn = o.save
+	if err := o.ResourceOptions.Complete(f, c, args); err != nil {
+		return err
+	}
+	o.ResourceOptions.Builder.InfoRetriever(RawRetriever{})
+
+	return nil
+}
+
+func (o *MigrateAPIStorageOptions) Validate() error {
+	if len(o.Mappings) == 0 {
+		return fmt.Errorf("at least one mapping argument must be specified: REGISTRY/NAME=REGISTRY/NAME")
+	}
+	return o.ResourceOptions.Validate()
+}
+
+func (o *MigrateAPIStorageOptions) Run() error {
+	return o.ResourceOptions.Visit(func(info *resource.Info) (migrate.Reporter, error) {
+		return o.transform(info.Object)
+	})
+}
+
+// save invokes the API to alter an object. The reporter passed to this method is the same returned by
+// the migration visitor method (for this type, transformAPIStorage). It should return an error
+// if the input type cannot be saved. It returns migrate.ErrRecalculate if migration should be re-run
+// on the provided object.
+func (o *MigrateAPIStorageOptions) save(info *resource.Info, reporter migrate.Reporter) error {
+	switch t := info.Object.(type) {
+	// TODO: add any custom mutations necessary
+	default:
+		info.Client.Put().
+			Resource(info.Mapping.Resource).
+			NamespaceIfScoped(info.Namespace, info.Mapping.Scope.Name() == meta.RESTScopeNameNamespace).
+			Name(info.Name).Body(info)
+		if _, err := resource.NewHelper(info.Client, info.Mapping).Replace(info.Namespace, info.Name, false, info.Object); err != nil {
+			return migrate.DefaultRetriable(info, err)
+		}
+		//return fmt.Errorf("resource %q does not have a save method implemented (%T)", info.Mapping.Resource, t)
+	}
+	return nil
+}
+
+// transform checks image references on the provided object and returns either a reporter (indicating
+// that the object was recognized and whether it was updated) or an error.
+func (o *MigrateAPIStorageOptions) transform(obj runtime.Object) (migrate.Reporter, error) {
+	fn := o.Mappings.MapReference
```

Comment:
It's not common?  Not sure what you're talking about exactly, but this
method is not generic.

On Thu, Jun 16, 2016 at 6:13 AM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/cmd/admin/migrate/storage/storage.go
> <https://github.com/openshift/origin/pull/9240#discussion_r67320184>:
>
> > +		info.Client.Put().
> > +			Resource(info.Mapping.Resource).
> > +			NamespaceIfScoped(info.Namespace, info.Mapping.Scope.Name() == meta.RESTScopeNameNamespace).
> > +			Name(info.Name).Body(info)
> > +		if _, err := resource.NewHelper(info.Client, info.Mapping).Replace(info.Namespace, info.Name, false, info.Object); err != nil {
> > +			return migrate.DefaultRetriable(info, err)
> > +		}
> > +		//return fmt.Errorf("resource %q does not have a save method implemented (%T)", info.Mapping.Resource, t)
> > +	}
> > +	return nil
> > +}
> > +
> > +// transform checks image references on the provided object and returns either a reporter (indicating
> > +// that the object was recognized and whether it was updated) or an error.
> > +func (o *MigrateAPIStorageOptions) transform(obj runtime.Object) (migrate.Reporter, error) {
> > +	fn := o.Mappings.MapReference
>
> How about moving those into some migration commons/utils instead of
> identical copy as in pkg/cmd/admin/migrate/images/images.go ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9240/files/eb1eb8a1f37fe62714438a3a805545a98b82daec..053e3598622ec47ce57195dacc03aedc08e4489f#r67320184>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p-7DdgqGnJupOYGdCmZfuf27TYLsks5qMSHPgaJpZM4IxnZA>
> .
>